osu.Game/Rulesets/Difficulty/Skills/Skill.cs

```diff
@@ -1,120 +1,70 @@
         /// </summary>
-        protected abstract double SkillMultiplier { get; }
+        protected IReadOnlyList<Mod> Mods => mods;
 
-        /// <summary>
-        /// Determines how quickly strain decays for the given skill.
-        /// For example a value of 0.15 indicates that strain decays to 15% of its original value in one second.
-        /// </summary>
-        protected abstract double StrainDecayBase { get; }
-
-        /// <summary>
-        /// The weight by which each strain value decays.
-        /// </summary>
-        protected virtual double DecayWeight => 0.9;
-
-        /// <summary>
-        /// <see cref="DifficultyHitObject"/>s that were processed previously. They can affect the strain values of the following objects.
-        /// </summary>
-        protected readonly LimitedCapacityStack<DifficultyHitObject> Previous = new LimitedCapacityStack<DifficultyHitObject>(2); // Contained objects not used yet
-
-        /// <summary>
-        /// The current strain level.
-        /// </summary>
-        protected double CurrentStrain { get; private set; } = 1;
+        private readonly Mod[] mods;
 
-        private double currentSectionPeak = 1; // We also keep track of the peak strain level in the current section.
-
-        private readonly List<double> strainPeaks = new List<double>();
+        protected Skill(Mod[] mods)
+        {
+            this.mods = mods;
+        }
 
         /// <summary>
-        /// Process a <see cref="DifficultyHitObject"/> and update current strain values accordingly.
+        /// Process a <see cref="DifficultyHitObject"/>.
         /// </summary>
+        /// <param name="current">The <see cref="DifficultyHitObject"/> to process.</param>
         public void Process(DifficultyHitObject current)
         {
-            CurrentStrain *= strainDecay(current.DeltaTime);
-            CurrentStrain += StrainValueOf(current) * SkillMultiplier;
+            // Preprocessing
+            RemoveExtraneousHistory(current);
 
-            currentSectionPeak = Math.Max(CurrentStrain, currentSectionPeak);
+            // Processing
+            Calculate(current);
 
+            // Postprocessing
             Previous.Push(current);
         }
 
         /// <summary>
-        /// Saves the current peak strain level to the list of strain peaks, which will be used to calculate an overall difficulty.
+        /// Remove objects from <see cref="Previous"/> that are no longer needed for calculations from the current object onwards.
         /// </summary>
-        public void SaveCurrentPeak()
+        /// <param name="current">The <see cref="DifficultyHitObject"/> to be processed.</param>
+        protected virtual void RemoveExtraneousHistory(DifficultyHitObject current)
         {
-            if (Previous.Count > 0)
-                strainPeaks.Add(currentSectionPeak);
+            // Default implementation to not retain objects
+            Previous.Clear();
```

Comment:
Would probably keep the previous implementation rather than just clearing like this. It makes `Previous` essentially useless unless overridden.

Reply:
That's the point though?
`Previous` _should be_ useless (in the sense that it doesn't store anything) if there is no requirement for it by the `Skill`.

Either way, with these changes the `Previous` collection is actually more efficient than the existing `LimitedCapacityStack` even for the same `Skill` implementations, since now we are only storing at most two objects in the collection at any given time, as opposed to before where it was a constant three, right?
Although that wasn't a flaw of the `LimitedCapacityStack` but instead that for whatever reason we left the capacity as three even though it was never used.

Reply:
I get what you mean though. For the base skill we have this flow of adding an object to the collection, just for it to be inevitably cleared out when processing the next object.
But this is a case where I think the performance benefit wouldn't be worth the slightly unintuitive code to handle this case.
My initial thought would be to have an empty virtual method for `RemoveExtraneousHistory` and a conditional in `Process` and the ctor to check for the special case where `PreviousCollectionSoftCapacity` is 0 and simply skip initialising or enqueuing to `Previous`.

Reply:
Empty override is fine, but the size checks are weird. If anything I'd also leave the addition to the consumer, the same way removal is currently. Maybe make it one method, even, after `Calculate()`. `UpdateHistory(DifficultyHitObject newObject)` or something?

Reply:
I'm not a super big fan of this approach, since the `Previous` stack is an idea handled by the `Skill` class, it makes sense for the `Skill` class to handle as much of the logic as it reasonably can.
I'm happy to make the change if you insist since it doesn't affect functionality, but I don't like that we are going to now have copy pasted code in all `Skill` implementations simply to accommodate for this edge case which is almost never relevant (since `Skill`s almost always need access to atleast the single prior object).

Reply:
> Maybe make it one method, even, after `Calculate()`. `UpdateHistory(DifficultyHitObject newObject)` or something?

This also can't be a single method, since removal of irrelevant objects needs to be done as a preprocessing step for a given object.
After adding the current object, we can't know for sure what the next objects requirements will be.

Reply:
The only way we can make it a single method is by passing the _previous_ object along to `Skill.Process` along with the current one, so that we can add the previous and remove the unnecessary all in one step.
But this now creates inconsistent `Skill` state between `Process` calls again which is something I want to avoid if at all possible.

Reply:
> since the `Previous` collection is an idea handled by the `Skill` class, it makes sense for the `Skill` class to handle as much of the logic as it reasonably can

It's not unreasonable to have someone i.e. want to ignore spinners entirely in a skill, or something along these lines. You could have a sane base implementation that adds the object to the queue and doesn't need to be touched in the 99% case.

Reply:
Ah yea that's true. I'm happy to make it the same as the `RemoveExtraneousHistory` then. Both can be empty in the base `Skill`, with proper logic in the `StrainSkill override`.
Changed in e61ae58ff1e771f50cfe6e7e2b5de85fba607b2e